### PR TITLE
[DNM] Interface definitions and initialization points for manipulators

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose

This PR is based on work done for a [direct manipulation prototype] (https://github.com/pboyer/Dynamo/blob/MFFY-Recharge/src/DynamoCore/Manipulation). It introduces interface definitions in `DynamoCoreWpf` to support the creation of direct manipulation manipulators as well as initialization of these manipulators in `DynamoView`.

The `ManipulatorDaemonInitializer` property in `DynamoViewModel` is instantiated when it is passed an object of type `IManipulatorDaemonInitializer` from Dynamo Studio. Manipulators can be created in `DynamoView` only if this object is initialized. The classes implementing these interfaces will reside in Dynamo Studio repo and I will be creating a separate PR for that. 

The `Selection_CollectionChanged` event handler in `DynamoView` creates a manipulator for the specific `NodeModel` type in the collection, if a manipulator is registered for that `NodeModel` type.

### Reviewers

@Benglin Reviewer 1

@pboyer Reviewer 2

### FYIs

@Randy-Ma

### CODE REVIEW

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 1]
```diff
@@ -647,6 +651,29 @@ void Selection_CollectionChanged(object sender, System.Collections.Specialized.N
...
+            {
+                manipulatorDaemon.KillAll();
+            }
+
+            if (e.NewItems == null)
+                return;
+
+            foreach (var nm in e.NewItems.OfType<NodeModel>())
+                manipulatorDaemon.CreateManipulator(nm, this);
         }
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: @Benglin @aparajit-pratap Nice, so this is where manipulators are created and destroyed.  They are "attached" to the node corresponding to different selection changes.  We might want to move this code to a specific method to clarify.

File: src/DynamoCoreWpf/Manipulation/Manipulator.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+using Autodesk.DesignScript.Interfaces;
+using Dynamo.Wpf.Interfaces;
+
+namespace Dynamo.Wpf.Manipulation
+{
+    public abstract class NodeManipulatorCreator : INodeManipulatorCreator
+    {
+        /// <summary>
+        /// Create a set of manipulators for the given node by inspecting its numeric inputs
+        /// </summary>
```
Comment (pboyer) [inline src/DynamoCoreWpf/Manipulation/Manipulator.cs]: This comment is misleading.

File: src/DynamoCoreWpf/Manipulation/DynamoManipulatorContext.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+using System.Text;
+using Dynamo.Controls;
+
+namespace Dynamo.Wpf.Manipulation
+{
+    public class DynamoManipulatorContext
+    {
+        internal DynamoManipulatorContext() { }
+
+        public DynamoView View { get; set; }
```
Comment (pboyer) [inline src/DynamoCoreWpf/Manipulation/DynamoManipulatorContext.cs]: We should not pass out the `DynamoView` to client code like this. We should make the property immutable by only expose `{ get; }`.  

My suggestion is to leave this property out until we have more specific API requirements.

Comment (pboyer) [inline src/DynamoCoreWpf/Manipulation/DynamoManipulatorContext.cs]: We're not *totally* sure that the `DynamoView` is the only context in which a manipulator would be created. I can imagine other contexts.

**Comment by pboyer** [comment]
I'd strongly consider your choice to put the manipulator stuff in the WPF part of the app. I would suggest making a separate project, `DynamoManipulation` and host the code there. Then, you put WPF specific code inside of `DynamoManipulatorWpf`. This model will make your code much more reusable and easier to test.

File: src/DynamoCoreWpf/Manipulation/DynamoManipulatorContext.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Dynamo.Controls;
+
+namespace Dynamo.Wpf.Manipulation
```
Comment (pboyer) [inline src/DynamoCoreWpf/Manipulation/DynamoManipulatorContext.cs]: Let's not put this in the WPF namespace unless we have to. I can very much see non-WPF apps wanting to create manipulators...

File: src/DynamoCoreWpf/Manipulation/Manipulator.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        public abstract void Dispose();
+
+        public abstract void Tessellate(IRenderPackage package, TessellationParameters parameters);
+    }
+
+    internal class CompositeManipulator : Manipulator
+    {
+        private readonly List<IManipulator> subManipulators;
+        public CompositeManipulator(List<IManipulator> manipulators)
```
Comment (pboyer) [inline src/DynamoCoreWpf/Manipulation/Manipulator.cs]: This could be generalized to `IEnumerable`. And then copied in the constructor.  That way, client code does not have a mutable copy of `subManipulators`.

File: src/DynamoCoreWpf/Manipulation/DynamoManipulatorContext.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class DynamoManipulatorContext
```
Comment (pboyer) [inline src/DynamoCoreWpf/Manipulation/DynamoManipulatorContext.cs]: Couldn't this be called simply `NodeManipulatorContext`? Or more simply, `ManipulatorContext`?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModelBranding.cs [hunk 7]
```diff
@@ -1,9 +0,0 @@
-﻿using Dynamo.Wpf.Interfaces;
-
-namespace Dynamo.ViewModels
-{
-    partial class DynamoViewModel 
-    {
-        public IBrandingResourceProvider BrandingResourceProvider { get; private set; }
-    }
-}
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModelBranding.cs]: I'm guessing this is an unrelated change...

**Comment by RodRecker** [comment]
I agree with Peter's comment about creating the DynamoManipulation project.

**Comment by aparajit-pratap** [comment]
@pboyer thanks for taking a look. By having `DynamoManipulation` as a separate project, are you suggesting `DynamoCoreWpf` to have a dependency on `DynamoManipulation` just like it has on `DynamoPackages`? Or do you suggest that I implement `IExtension` for manipulator classes and completely remove the dependency of `DynamoCore`/`DynamoCoreWpf` from `DynamoManipulation`? The reason why everything is in `DynamoCoreWpf` currently is because the manipulators are initialized in `DynamoView`.

**Comment by pboyer** [comment]
@aparajit-pratap Take a look at [`IViewExtension`](https://github.com/DynamoDS/Dynamo/blob/9de002f8743d57e1d35698483b2c26830b5f6cd6/src/DynamoCoreWpf/Extensions/IViewExtension.cs)

It actually supplies hooks into the application lifetime. This should allow you to remove it completely.

**Comment by aparajit-pratap** [comment]
@pboyer I hear you. I did take a look at `IExtension` and `IViewExtension`. I think it's the way to go forward although I feel I should to do this in phases. To begin with I wish to get something working quickly and have something to further refine and build on top of. However I will eventually make this project as an `IViewExtension`. Thanks.
